src/lib/HST/CoreAlgorithm.hs

```diff
@@ -78,41 +47,43 @@ firstPat = head . fst
 -------------------------------------------------------------------------------
 -- Wadler's Algorithm                                                        --
 -------------------------------------------------------------------------------
-
 -- | The default error expression to insert for pattern matching failures.
 defaultErrorExp :: S.Exp a
-defaultErrorExp =
-  S.Var S.NoSrcSpan (S.UnQual S.NoSrcSpan (S.Ident S.NoSrcSpan "undefined"))
+defaultErrorExp = S.Var S.NoSrcSpan
+  (S.UnQual S.NoSrcSpan (S.Ident S.NoSrcSpan "undefined"))
 
 -- | Compiles the given equations of a function declaration to a single
 --   expression that performs explicit pattern matching using @case@
 --   expressions.
 --
 --   All equations must have the same number of patterns as the given list
 --   of fresh variable patterns.
-match
-  :: (Members '[Env a, Fresh, GetOpt, Report] r, S.EqAST a)
-  => [S.Pat a] -- ^ Fresh variable patterns.
-  -> [Eqs a]   -- ^ The equations of the function declaration.
-  -> S.Exp a   -- ^ The error expression for pattern-matching failures.
-  -> Sem r (S.Exp a)
-match [] (([], e) : _) _  = return e  -- Rule 3a: All patterns matched.
-match [] []            er = return er -- Rule 3b: Pattern-matching failure.
+match :: (Members '[Env a, Fresh, GetOpt, Report] r, S.EqAST a)
+      => [S.Pat a] -- ^ Fresh variable patterns.
+      -> [Eqs a]   -- ^ The equations of the function declaration.
+      -> S.Exp a   -- ^ The error expression for pattern-matching failures.
+      -> Sem r (S.Exp a)
+match [] (([], e) : _) _ = return e  -- Rule 3a: All patterns matched.
+match [] [] er = return er -- Rule 3b: Pattern-matching failure.
 match vars@(x : xs) eqs er
-  | -- Rule 1: Pattern list of all equations starts with a variable pattern.
+  |
+    -- Rule 1: Pattern list of all equations starts with a variable pattern.
     allVarPats eqs = do
-    eqs' <- mapM (substVars x) eqs
-    match xs eqs' er
-  | -- Rule 2: Pattern lists of all equations starts with a constructor pattern.
+      eqs' <- mapM (substVars x) eqs
+      match xs eqs' er
+  |
+    -- Rule 2: Pattern lists of all equations starts with a constructor pattern.
     allConPats eqs = makeRhs x xs eqs er
-  | -- Rule 4: Pattern lists of some equations start with a variable pattern
+  |
+    -- Rule 4: Pattern lists of some equations start with a variable pattern
```

Comment:
Maybe it just looks like the comments are in the wrong order due to the diff. The actual code looks right to me.

```haskell
match vars@(x : xs) eqs er
  |
    -- Rule 1: Pattern list of all equations starts with a variable pattern.
    allVarPats eqs = do
      eqs' <- mapM (substVars x) eqs
      match xs eqs' er
  |
    -- Rule 2: Pattern lists of all equations starts with a constructor pattern.
    allConPats eqs = makeRhs x xs eqs er
  |
    -- Rule 4: Pattern lists of some equations start with a variable pattern
    -- and others start with a constructor pattern.
    --
    -- TODO This is probably causing the infinite loops when there are
    --      unsupported patterns.
    otherwise = createRekMatch vars er (groupByFirstPatType eqs)
```